src/main/scala/dsls/regex/RegularExpression.scala

```diff
@@ -10,9 +13,50 @@ package dsls.regex
 /** The top of a class hierarchy that encodes regular expressions. */
 abstract class RegularExpression {
   /** returns true if the given string matches this regular expression */
+  
   def matches(string: String) = RegexMatcher.matches(string, this)
+  
+  // || represents the Union of this, which is of type RegularExpression, 
+  // with "other", which is also of type RegularExpression
+  def || (other: RegularExpression) = Union(this, other)
+  
+  // || represents the Concat of this, which is of type RegularExpression, 
+  // with "other", which is also of type RegularExpression
+  def ~ (other: RegularExpression) = Concat(this, other) 
+  
+  def <*> = Star (this)
+  
+  // another way to do it is def <+> = this ~ (this <*>)
+  def <+> = Concat (this, Star(this)) 
+  
+
+  def apply(n: Int): RegularExpression = {
+    n match {
+        case 0 => EPSILON // base case 
+        case n if n >= 1 => this ~ apply(n-1) // concat "this" with what "apply" recursively returns 
+        // anticipate a possible input error and tries to give an informative error message 
+        case _ => error("Illegal argument: n cannot be negative.") // anticipate a possible input error and tries
+      }
+    }
+}
+  
+object RegularExpression {
+  
+    // converts from a character to a literal 
+   implicit def Char2Lit (c : Char): Literal = new Literal (c)
+ 
+   /* Converts a string to a RegularExpression using concatenation */
+    // recursively concatenates each subsequent character in a string until you've gone through entire string  
+	  implicit def String2RegExpr(str: String): RegularExpression = {
+      if (str.isEmpty)
+           EPSILON
+      else
+           // recursively concat first elt of string to rest 
+           str.head ~ (String2RegExpr(str.tail))
+    }
```

Comment:
Nice! If you are into higher-order functions, you can refactor this using a right fold:
```scala
implicit def string2RegExpr(str: String): RegularExpression = str.foldRight(EPSILON: RegularExpression) ({(l, acc) => l ~ acc})
```